src/charm.py

```diff
@@ -66,273 +63,17 @@ def __init__(self, *args):
-
-        if password == auth.get_password(DEFAULT_ADMIN_USERNAME):
-            event.log("The old and new passwords are equal.")
-            event.set_results({"password": password})
-            return
-
-        auth.set_password(DEFAULT_ADMIN_USERNAME, password)
-
-        event.set_results({"password": password})
-
-    @property
-    def s3_connection_info(self) -> Optional[S3ConnectionInfo]:
-        """Parse a S3ConnectionInfo object from relation data."""
-        # If the relation is not yet available, return None
-        if not self.s3_requirer.relations:
-            return None
-
-        raw_info = self.s3_requirer.get_s3_connection_info()
-
-        if any(key not in raw_info for key in ["endpoint", "access-key", "secret-key"]):
-            return None
-
-        return S3ConnectionInfo(
-            endpoint=raw_info.get("endpoint"),
-            access_key=raw_info.get("access-key"),
-            secret_key=raw_info.get("secret-key"),
-            path=raw_info.get("path"),
-            bucket=raw_info.get("bucket"),
-        )
-
-    @property
-    def metastore_db_connection_info(self) -> Optional[DatabaseConnectionInfo]:
-        """Parse a DatabaseConnectionInfo object from metastore_db relation data."""
-        # If the relation is not yet available, return None
-        if not self.metastore_db.relations:
-            return None
-
-        raw_info = self.metastore_db.fetch_relation_data()
-        for data in raw_info.values():
-            if any(key not in data for key in ["endpoints", "username", "password"]):
-                continue
-            return DatabaseConnectionInfo(
-                endpoint=data["endpoints"],
-                username=data["username"],
-                password=data["password"],
-                dbname=METASTORE_DATABASE_NAME,
-            )
-        return None
-
-    @property
-    def auth_db_connection_info(self) -> Optional[DatabaseConnectionInfo]:
-        """Parse a DatabaseConnectionInfo object from metastore_db relation data."""
-        # If the relation is not yet available, return None
-        if not self.auth_db.relations:
-            return None
-
-        raw_info = self.auth_db.fetch_relation_data()
-        for data in raw_info.values():
-            if any(key not in data for key in ["endpoints", "username", "password"]):
-                continue
-            hostname, port = data["endpoints"].split(":")
-            return DatabaseConnectionInfo(
-                endpoint=hostname,
-                username=data["username"],
-                password=data["password"],
-                dbname=AUTHENTICATION_DATABASE_NAME,
-            )
-        return None
-
-    def _on_s3_credential_changed(self, _: CredentialsChangedEvent):
-        """Handle the `CredentialsChangedEvent` event from S3 integrator."""
-        logger.info("S3 credentials changed")
-        self.update_service()
+        self.s3 = S3Requirer(self, S3_INTEGRATOR_REL)
```

Comment:
This was moved here for consistency earlier because I had to move `DatabaseRequires` to the top level earlier. However, since I am going to split `DatabaseRequires` to `*Data` and `*EventHandler` implementations separately, I'll move this down the line to event handlers.